spring-cloud-zookeeper-discovery/src/main/java/org/springframework/cloud/zookeeper/discovery/configclient/ZookeeperConfigServerBootstrapper.java

```diff
@@ -162,35 +130,63 @@ public List<ServiceInstance> apply(String serviceId) {
 			return apply(serviceId, null, null, null);
 		}
 
+		private static <T> Supplier<T> supplier(BootstrapContext context, Class<T> type) {
+			if (context.isRegistered(type)) {
+				T instance = context.get(type);
+				return () -> instance;
+			}
+			else {
+				return () -> null;
+			}
+		}
+
 		@Override
 		public List<ServiceInstance> apply(String serviceId, Binder binder, BindHandler bindHandler, Log log) {
 			if (binder == null || !isEnabled(binder)) {
 				return Collections.emptyList();
 			}
-
-			ZookeeperProperties properties = context.getOrElse(ZookeeperProperties.class, binder.bind(ZookeeperProperties.PREFIX, Bindable.of(ZookeeperProperties.class))
-					.orElse(new ZookeeperProperties()));
-			RetryPolicy retryPolicy = context.getOrElse(RetryPolicy.class, new ExponentialBackoffRetry(properties.getBaseSleepTimeMs(), properties.getMaxRetries(),
-					properties.getMaxSleepMs()));
-			try {
-				CuratorFramework curatorFramework = context.getOrElse(CuratorFramework.class, CuratorFactory.curatorFramework(properties, retryPolicy, Stream::of,
-						() -> null, () -> null));
-				InstanceSerializer<ZookeeperInstance> serializer = context.getOrElse(InstanceSerializer.class, new JsonInstanceSerializer<>(ZookeeperInstance.class));
-				ZookeeperDiscoveryProperties discoveryProperties = context.getOrElse(ZookeeperDiscoveryProperties.class, binder.bind(ZookeeperDiscoveryProperties.PREFIX, Bindable
-								.of(ZookeeperDiscoveryProperties.class), bindHandler)
-						.orElseGet(() -> new ZookeeperDiscoveryProperties(new InetUtils(new InetUtilsProperties()))));
-				ServiceDiscoveryCustomizer customizer = context.getOrElse(ServiceDiscoveryCustomizer.class, new DefaultServiceDiscoveryCustomizer(curatorFramework, discoveryProperties, serializer));
-				ServiceDiscovery<ZookeeperInstance> serviceDiscovery = customizer.customize(ServiceDiscoveryBuilder.builder(ZookeeperInstance.class));
-				ZookeeperDependencies dependencies = context.getOrElse(ZookeeperDependencies.class, binder.bind(ZookeeperDependencies.PREFIX, Bindable
-								.of(ZookeeperDependencies.class), bindHandler)
-						.orElseGet(ZookeeperDependencies::new));
-				return new ZookeeperDiscoveryClient(serviceDiscovery, dependencies, discoveryProperties).getInstances(serviceId);
-			}
-			catch (Exception e) {
-				log.warn("Error fetching config server instance from Zookeeper", e);
-				return Collections.emptyList();
+			if (this.discoveryClient == null) {
+				ZookeeperProperties properties = context.getOrElseSupply(ZookeeperProperties.class, () -> binder.bind(ZookeeperProperties.PREFIX, Bindable.of(ZookeeperProperties.class))
```

Comment:
I tried applying this `ZookeeperFunction` fix in my test project, but it didn't quite help in the end. The problem was that my customization code, that registers my `ZookeeperProperties` instance, had the same `isEnabled` condition as Spring itself, because I wanted to have the same activation behavior. But this creates the same problem with the `Binder`; I would need access to the "more knowledgeable" one when I register my custom beans.

Reply:
I think you can safely assume at the point where the beans are being requested from the bootstrap context that we have already done the isEnabled check (or we wouldn't be requesting them) so you shouldn't have to repeat that check for your custom beans

Reply:
Hmm, I guess that sounds reasonable. However I was simply copying that approach from `ZookeeperConfigServerBootstrapper.initialize` and [`CuratorFactory.registerCurator`](https://github.com/spring-cloud/spring-cloud-zookeeper/blob/v4.1.0/spring-cloud-zookeeper-core/src/main/java/org/springframework/cloud/zookeeper/CuratorFactory.java#L94-L106) – should the `isEnabled` guard be removed from all of the bootstrap context bean registrations? I'm trying to be as consistent with the framework's internals as possible.